test/suites/integration/metrics_test.go

```diff
@@ -0,0 +1,77 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration_test
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter-provider-aws/test/pkg/environment/common"
+)
+
+var _ = Describe("Metrics", func() {
+	It("should expose karpenter_cloudprovider_instance_type_offering_price_estimate metrics", func() {
+		env.ExpectCreated(nodeClass, nodePool)
+		out, err := env.EC2API.DescribeAvailabilityZones(env.Context, &ec2.DescribeAvailabilityZonesInput{})
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(func(g Gomega) {
+			defer GinkgoRecover()
+			podMetrics := env.ExpectPodMetrics()
+			priceMetricCount := lo.CountBy(podMetrics, func(p common.PrometheusMetric) bool {
+				return p.Name == "karpenter_cloudprovider_instance_type_offering_price_estimate"
+			})
+			nonZeroPriceMetricCount := lo.CountBy(podMetrics, func(p common.PrometheusMetric) bool {
+				return p.Name == "karpenter_cloudprovider_instance_type_offering_price_estimate" &&
+					p.Value > 0
+			})
+			// Assume that we find at least 700 instance types in a region
+			// with only 600 of them available for spot
+			expectedCount := 700 * len(out.AvailabilityZones) * 2
+			expectedNonZeroCount := 700*len(out.AvailabilityZones) + 600*len(out.AvailabilityZones)
+			g.Expect(priceMetricCount).To(BeNumerically(">", expectedCount))
+			g.Expect(nonZeroPriceMetricCount).To(BeNumerically(">", expectedNonZeroCount))
+		}, time.Minute, time.Second*5).Should(Succeed())
+
+	})
+	It("should expose karpenter_cloudprovider_instance_type_offering_available metrics", func() {
+		env.ExpectCreated(nodeClass, nodePool)
+		// Availability only has non-zero values for the subnets that we support
+		subnets := env.GetSubnets(map[string]string{"karpenter.sh/discovery": env.ClusterName})
```

Comment:
Should we include filtering to ensure this only counts one subnet per AZ? IIUC this would breakdown in my local environment (and probably others) since I have a public and private subnet per AZ.

Reply:
Good catch

Reply:
env.GetSubnets actually returns a map with the keys being the AZs where we have subnets, so this is actually already only capturing AZ count